### PR TITLE
Allow specifying build_root_image in the repository

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR allows to specify the `build_root_image` inside the repository by setting `build_root.from_repository: true`. If this is set, we will:
* Enable cloning on all prowjobs that are based off of that ci-operator config (We might optimize this later on to only do so when we actually have that image in our graph, but that requires some bigger refacotoring of the ci-operator and prowgen)
* Read the imagestream from a file named `.build_root_image` in the repo

/assign @stevekuznetsov 
/cc @marun 

/label tide/merge-method-squash

### CODE REVIEW

File: pkg/api/config.go
Comment (stevekuznetsov) [inline pkg/api/config.go:276]: this is missing your new option

File: pkg/api/types.go
Comment (stevekuznetsov) [inline pkg/api/types.go:311]: Could we go with a `.ci-operator.yaml` or something to future-proof this?

File: pkg/defaults/defaults.go
Comment (stevekuznetsov) [inline pkg/defaults/defaults.go:609]: Why can't we expect structured data here in normal name/namespace/tag format like everywhere else in config?

File: pkg/prowgen/prowgen.go
Comment (stevekuznetsov) [inline pkg/prowgen/prowgen.go:131]: Is it worth it to try to make this shallow, too, to save on time?

**Review by stevekuznetsov — COMMENTED** [review]
Minor nits

File: pkg/api/types.go
Comment (alvaroaleman) [reply pkg/api/types.go:311]: Sure, so make it a struct (which for now will only have this one option)?

**Review by petr-muller — COMMENTED** [review]
LGTM modulo what Steve said.

File: pkg/prowgen/prowgen.go
Comment (alvaroaleman) [reply pkg/prowgen/prowgen.go:131]: I don't see an option to enable shallow cloning in the decoration utils

File: pkg/defaults/defaults.go
Comment (alvaroaleman) [reply pkg/defaults/defaults.go:610]: done

File: pkg/api/types.go
Comment (alvaroaleman) [reply pkg/api/types.go]: done

**Comment by sttts** [comment]
/hold

Why do we need another source other than Dockerfile?

**Comment by sttts** [comment]
And before unhold, this MUST go into an enhancement. Document what CI does.

**Comment by tnozicka** [comment]
I support the `.build_root_image` file. Dockefile FROM and build _root images are different.

This should also update the doc here https://github.com/openshift/ci-tools/blob/master/CONFIGURATION.md#build_root

**Comment by sttts** [comment]
Questions:
- How many non-standard build_root_images are used in which repos?
- What is the standard build_root_image? Where is its source code?
- How is the standard build_root_image derived from the Dockerfile FROM image? Are they linked? They must IMO.
- Which things-to-version are in the build images? protoc? Other generators?
- How is the standard build_root_image linked to the Kube version used by the control plane?

**Comment by alvaroaleman** [comment]
This PR doesn't change anything about what build_root you can or can not use (I believe there are no restrictions on this), just where to configure it, in order to allow for an atomic change that changes both code and image.

The build_root is used to as the image to clone the code into. The result of that in turn is used as the image to execute the `binary_build_commands`, `test_binary_build_commands` and `rpm_build_commands`.

The standard images used for the build_root are the `openshift/release:golang-$VERSION` ones, their dockerfiles are here: https://github.com/openshift/release/tree/b2770007d21ea33f7ac1ea2ad780904e127448b9/projects/origin-release

**Review by stevekuznetsov — APPROVED** [review]

**Comment by stevekuznetsov** [comment]
/hold

Let's get the documentation page updated with this feature and I think the implementation here is merge-able.

**Comment by alvaroaleman** [comment]
Docs are at https://server-ci-tools-1251.apps.ci.l2s4.p1.openshiftapps.com/help/ci-operator

**Comment by stevekuznetsov** [comment]
/lgtm
/hold cancel

LGTM!

**Comment by sttts** [comment]
This PR is definitely the right direction. I have not seen an answer for how we guarantee that we won't test using different base image than what is used in the Dockerfile. We will probably build something into build-machinery as CI is not willing to build or talk about a machanism that guarantees that from the start.

**Pull request merged**
